src/main/java/uk/ac/ebi/atlas/search/metadata/MetadataSearchDao.java

```diff
@@ -0,0 +1,92 @@
+    }
+
+
+    private String getInferredCellTypeForCellId(String experimentAccession, String cellId) {
+        var characteristicFields = "inferred_cell_type";
+        var queryBuilder =
+                new SolrQueryBuilder<SingleCellAnalyticsCollectionProxy>()
+                        .addQueryFieldByTerm(EXPERIMENT_ACCESSION, experimentAccession)
+                        .addQueryFieldByTerm(CELL_ID, cellId)
+                        .addQueryFieldByTerm(CHARACTERISTIC_NAME, characteristicFields);
+        var results = this.singleCellAnalyticsCollectionProxy.query(queryBuilder).getResults();
+
+        var inferredCellType = "";
+        try {
+            inferredCellType = results
+                    .stream()
+                    .map(entry -> entry.get(CHARACTERISTIC_VALUE.name()).toString())
+                    .collect(Collectors.toList()).get(0);
+        } catch (Exception e) {
+            LOGGER.debug("Invalid cell id: {}", cellId);
+        }
+        return inferredCellType;
+    }
+
+    // Given a set of characteristic name (organism part) and characteristic value (pancreas),
+    // this method retrieves the value of that metadata for list of cell IDs.
+    @Cacheable(cacheNames = "cellIdsMetadata", key = "{#characteristicName, #characteristicValue}")
+    public Map<String, Map<String, String>> getCellTypeMetadata(String characteristicName,
+                                                                String characteristicValue,
+                                                                String experimentAccession) {
+
+        var queryBuilder =
+                new SolrQueryBuilder<SingleCellAnalyticsCollectionProxy>()
+                        .addQueryFieldByTerm(EXPERIMENT_ACCESSION, experimentAccession)
+                        .addQueryFieldByTerm(CHARACTERISTIC_NAME, characteristicName)
+                        .addQueryFieldByTerm(CHARACTERISTIC_VALUE, characteristicValue);
+
+        var results = this.singleCellAnalyticsCollectionProxy.query(queryBuilder).getResults();
+
+        return results
+                .stream()
+                .collect(groupingBy(solrDocument -> (String) solrDocument.getFieldValue(EXPERIMENT_ACCESSION.name())))
+                .entrySet()
+                .stream()
+                .collect(
+                        toMap(
+                                Map.Entry::getKey,
+                                entry -> entry.getValue()
+                                        .stream()
+                                        .collect(
+                                                toMap(
+                                                        cell -> cell.get(CELL_ID.name()).toString(),
+                                                        cell -> getInferredCellTypeForCellId(
+                                                                entry.getKey(),
+                                                                cell.get(CELL_ID.name()).toString()
```

Comment:
As we are calling 'getInferredCellTypeForCellId' method for each cell id. Would it be possible to use Streaming Expressions instead of a normal Solr query?